runtime/lua/vim/lsp/util.lua

```diff
@@ -660,8 +663,19 @@ do
       local start = diagnostic.range.start
       local finish = diagnostic.range["end"]
 
+      local hl = underline_highlight_name
+      if diagnostic.severity == protocol.DiagnosticSeverity.Error then
+          hl = underline_highlight_name..'Error'
+      elseif diagnostic.severity == protocol.DiagnosticSeverity.Warning then
+          hl = underline_highlight_name..'Warning'
+      elseif diagnostic.severity == protocol.DiagnosticSeverity.Information then
+          hl = underline_highlight_name..'Information'
+      elseif diagnostic.severity == protocol.DiagnosticSeverity.Hint then
+          hl = underline_highlight_name..'Hint'
+      end
+
       -- TODO care about encoding here since this is in byte index?
-      highlight_range(bufnr, diagnostic_ns, underline_highlight_name,
+      highlight_range(bufnr, diagnostic_ns, hl,
```

Comment:
the if-else verbosity can be avoided by using a map (which may later be useful as part of the `protocol` module if it doesn't already exist...


```
local hlmap = {
  [protocol.DiagnosticSeverity.Error]='Error',
  [protocol.DiagnosticSeverity.Warning]='Warning',
  [protocol.DiagnosticSeverity.Information]='Information',
  [protocol.DiagnosticSeverity.Hint]='Hint',
}
highlight_range(bufnr, diagnostic_ns,
  underline_highlight_name..hlmap[diagnostic.severity],
    {start.line, start.character},
    {finish.line, finish.character}
)
```